doc/innoextract.1
.\" Manpage for innoextract.
.\" Contact daniel@constexpr.org to correct errors or typos.
.TH innoextract 1 "2016-03-02" "1.7-dev"
.SH NAME
innoextract - tool to extract installers created by Inno Setup
.SH SYNOPSIS
.B innoextract
.RB [ \-\-extract ]
.RB [ \-\-lowercase ]
[options] [\fB\-\-\fP] \fIinstallers\fP ...

\fBinnoextract \-\-list\fP [options] [\fB\-\-\fP] \fIinstallers\fP ...

\fBinnoextract \-\-test\fP [options] [\fB\-\-\fP] \fIinstallers\fP ...
.SH DESCRIPTION
\fBinnoextract\fP is a tool that can extract installer executables created by Inno Setup.
.PP
\fBinnoextract\fP will extract files from installers specified on the command line.
.PP
To extract a multi-part installer with external data files, only the executable (.exe) file needs to be given as an argument to \fBinnoextract\fP.
.SH OPTIONS SUMMARY
.PP
Here is a short summary of the options available in innoextract. Please refer to the detailed  documentation below for a complete description.
.TP
.B Generic options:
.nf
 \-h \-\-help               Show supported options
 \-v \-\-version            Print version information
    \-\-license            Show license information
.fi
.TP
.B Actions:
.nf
 \-t \-\-test               Only verify checksums, don't write anything
 \-e \-\-extract            Extract files (default action)
 \-l \-\-list               Only list files, don't write anything
    \-\-list\-sizes         List file sizes
    \-\-list\-checksums     List file checksums
 \-i \-\-info               Print information about the installer
    \-\-list\-languages     List languages supported by the installer
    \-\-gog\-game\-id        Determine the GOG.com game ID for this installer
    \-\-show\-password      Show password check information
    \-\-check\-password     Abort if the password is incorrect
 \-V \-\-data\-version       Only print the data version
.fi
.TP
.B Modifiers:
.nf
    \-\-collisions \fIACTION\fP  How to handle duplicate files
    \-\-default\-language   Default language for renaming
    \-\-dump               Dump contents without converting filenames
 \-L \-\-lowercase          Convert extracted filenames to lower-case
 \-T \-\-timestamps \fITZ\fP      Timezone for file times or "local" or "none"
 \-d \-\-output\-dir \fIDIR\fP     Extract files into the given directory
 \-P \-\-password \fIPASSWORD\fP  Password for encrypted files
    \-\-password\-file \fIFILE\fP File to load password from
 \-g \-\-gog                Process additional archives from GOG.com installers
    \-\-no\-gog\-galaxy      Don't re-assemble GOG Galaxy file parts
 \-n \-\-no\-extract\-unknown Don't extract unknown Inno Setup versions
.fi
.TP
.B Filters:
.nf
 \-m \-\-exclude\-temp       Don't extract temporary files
    \-\-language \fILANG\fP      Extract only files for this language
    \-\-language\-only      Only extract language-specific files
 \-I \-\-include \fIEXPR\fP       Extract only files that match this path
.fi
.TP
.B Display options:
.nf
 \-q \-\-quiet              Output less information
 \-s \-\-silent             Output only error/warning information
    \-\-no\-warn\-unused     Don't warn on unused \fI.bin\fP files
 \-c \-\-color[=\fIENABLE\fP]     Enable/disable color output
 \-p \-\-progress[=\fIENABLE\fP]  Enable/disable the progress bar
.fi
.SH OPTIONS
.TP
\fB--\fP
Treat all arguments after this one as files, even if they begin with a dash.
.TP
\fB\-\-check\-password\fB
Abort processing if the password provided using the \fB\-\-password\fP or \fB\-\-password\-file\fP option does not match the checksum stored in the installer.

The password checksum used for this check can be retrieved using the \fB\-\-show\-password\fP option.
.TP
\fB\-\-collisions\fP \fIACTION\fP
Inno Setup installers can contain duplicate files with the same name. This option tells innoextract what to do when such a collisions is encountered. Valid actions are:

.RS
.TP
"\fBoverwrite\fP"
Extract only one of the colliding files. The choice is done similar to how Inno Setup overwrites files during installation. This is the default.
.TP
"\fBrename\fP"
Rename files that would be overwritten using the "\fBoverwrite\fP" action by appending a suffix comprised of the file's language, the component it belongs to and/or a number to make the filename unique. The language suffix (if applicable) is also appended to the \fIdefault\fP file that would have been extracted with the "\fBoverwrite\fP" action.
.TP
"\fBrename-all\fP"
Rename all colliding files by appending a suffix comprised of the file's language, the component it belongs to and/or a number to make the filename unique. The complete suffix is appended to both files that would have been overwritten using the "\fBoverwrite\fP" action and to those that would have overwritten other files.
.TP
"\fBerror\fP"
Exit when a collision is detected.
.RE
.IP
.B Rename rules:

1. If the \fBcomponent\fP is not the same for all files in the collision set (all files with the same filename), "\fB#\fP" (without quotes) followed by the component id is appended to all files that are specific to a single component.

2. If the \fBlanguage\fP is not the same for all files in the collision set, "\fB@\fP" (without quotes) followed by the language id is appended to all files that are specific to a single component unless that language matches the default language specified by the \fB--default-language\fP. While the suffix is omitted for the default language, no numbered suffix is added in it's place unless needed to make the filename unique.

3. If no suffix was added by the previous steps, or if the filename is not yet unique, "\fB$\fP" (without quotes) followed by the lowest integer (starting at 0) to make the filename unique is appended.

With the "\fBrename\fP" action, steps 1 and 3 are only applied to files that would have been overwritten by the "\fBoverwrite\fP" action while "\fBrename-all\fP" applies them to all files in the collision set.
.TP
\fB\-c\fP, \fB\-\-color\fP[=\fIENABLE\fP]
By default
.B innoextract
will try to detect if the terminal supports shell escape codes and enable or disable color output accordingly. Specifically, colors will be enabled if both \fBstdout\fP and \fBstderr\fP point to a TTY and the \fBTERM\fP environment variable is not set to "\fBdumb\fP". Pass \fB1\fP or \fBtrue\fP to \fB\-\-color\fP to force color output. Pass \fB0\fP or \fBfalse\fP to never output color codes.
.TP
\fB\-V\FP, \fB\-\-data\-version\fP
Print the Inno Setup data version of the installer and exit immediately.

The version printed using this option is the one stored in the setup file and can differ from the version printed with other actions as the stored data version is not always correct.

This option can be used to determine if a file is an Inno Setup installer without loading any compressed headers.

This option cannot be combined with any other action.
.TP
\fB\-\-default\-language\fP \fILANG\fP
Set a language as the default.

With \fB\-\-collisions\=overwrite\fP (the default) this will change the choice of which file to keep to always prefer the given language. In effect, \fB\-\-default\-language\fP behaves almost like \fB\-\-language\fP, except that files are extracted for all languages if they have different names.

When using the \fB\-\-collisions\=rename\fP option, \fB\-\-default\-language\fP chooses a language for which the files should keep the original name if possible.
.TP
\fB\-\-dump\fP
Don't convert Windows paths to UNIX paths and don't substitute constants in paths.

When combining \fB\-\-dump\fP with \fB\-\-extract\fP innoextract will \fInot\fP ensure that the paths don't point outside the destination directory. Use this option with caution when handling untrusted files.
.TP
\fB\-m\fP, \fB\-\-exclude\-temp\fP
Don't extract files that would have been deleted at the end of the install process. Such files are marked with [temp] in the file listing.

This option takes precedence over \fB\-\-include\fP and \fB\-\-language\fP: temporary files are never extracted when using the \fB\-\-exclude\-temp\fP, even if they match the selected language or include expressions.
.TP
\fB\-e\fP, \fB\-\-extract\fP
Extract all files to the current directory. This action is enabled by default, unless one or more of the \fB\-\-list\fP, \fB\-\-list\-sizes\fP, \fB\-\-list\-checksums\fP, \fB\-\-test\fP, \fB\-\-list\-languages\fP, \fB\-\-gog\-game\-id\fP, \fB\-\-show\-password\fP or \fB\-\-check\-password\fP actions are specified.

Be default innoextract will continue after encountering file checksum errors. The \fB\-\-extract\fP option can be combined with \fB\-\-test\fP to abort on checksum errors.
.TP
\fB\-n\fP, \fB\-\-no\-extract\-unknown\fP
By default innoextract will try to extract installers with and unknown Inno Setup data version by treating it as the closest known version. This option tells innoextract to aborts instead.
.TP
\fB\-g\fP, \fB\-\-gog\fP
Try to process additional .bin files that have the same basename as the setup but are not actually part of the Inno Setup installer. This is the case for newer multi-part GOG.com installers where these .bin files are RAR archives, potential encrypted with the MD5 checksum of the game ID (see the \fB\-\-gog\-game\-id\fP option).

Extracting these RAR archives requires rar, unrar or lsar/unar command-line utilities to be in the PATH.

The \fB\-\-list\fP, \fB\-\-test\fP, \fB\-\-extract\fP and \fB\-\-output\-dir\fP options are passed along to unrar/unar, but other options may be ignored for the RAR files. For multi-part RAR archives, the \fB\-\-test\fP requires a writable output directory for temporary files.

Note that is option is geared towards GOG.com installers. Other installers may come be bundled with different extraneous \fI.bin\fP which this option might not be able to handle.

This option also forces re-assembly of GOG Galaxy file parts. See the \fB\-\-no\-gog\-galaxy\fP for details.
.TP
\fB\-\-no\-gog\-galaxy\fP
Some GOG.com installers contain files in GOG Galaxy format (split into multiple parts that are individually compressed) which are re-assembled using post-install scripts. By default \fBinnoextract\fP will try to re\-assemble such files if it detects a GOG.com installer. This option disables that.

GOG.com installers are detected using the publisher and URL fields in the setup headers. Use the \fB\-\-gog\fP option to force reassembly for all installers.
.TP
\fB\-\-gog\-game\-id\fP
Determine the ID used by GOG.com for the game contained in this installer. This will only work with Galaxy-ready GOG.com installers.

This option can be combined with \fB\-\-silent\fP to print only the game ID without additional syntax that would make consumption by other scripts harder.

The \fB\-\-gog\-game\-id\fP action can be combined with \fB\-\-list\fP, \fB\-\-test\fP, \fB\-\-extract\fP and/or \fB\-\-list\-languages\fP. If \fB\-\-silent\fP and \fB\-\-gog\-game\-id\fP are combined with \fB\-\-list\fP and/or \fB\-\-list\-languages\fP, the game ID (or an empty line) will be printed on it's own line before the file list but after the language list.

For newer multi-part GOG.com installers the \fI.bin\fP files are not part of the Inno Setup installer but instead are RAR archives. Some of these RAR files are encrypted, with the password being the MD5 checksum of the game ID:

  \fBinnoextract \-\-gog\-game\-id --silent\fP \fIsetup_....exe\fP | \fBmd5sum\fP | \fBcut \-d\fP ' ' \fB\-f\fP 1
.TP
\fB\-h\fP, \fB\-\-help\fP
Show a list of the supported options.
.TP
\fB\-I\fP, \fB\-\-include\fP \fIEXPR\fP
If this option is specified, innoextract will only process files whose path matches \fIEXPR\fP. The expression can be either a single path component (a file or directory name) or a series of successive path components joined by the OS path separator (\\ on Windows, / elsewhere).

The expression is always matched against one or more full path components. Filtering by parts of filenames is currently not supported. Matching is done case-insensitively.

\fIEXPR\fP may contain one leading path separator, in which case the rest of the expression is matched against the start of the path. Otherwise, the expression is matched against any part of the path.

The \fB\-\-include\fP option may be repeated in order allow files matching against one of multiple patterns. If \fB\-\-include\fP is not used, all files are processed.
.TP
\fB\-i\fP \fB\-\-info\fP
This is a convenience option to enable all actions that print information about the installer.

Scrips should not rely on the output format with this option and should instead enable the individual actions instead.

Currently this option enables \fB\-\-list\-languages\fP, \fB\-\-gog\-game\-id\fP and \fB\-\-show\-password\fP.
.TP
\fB\-\-language\fP \fILANG\fP
Extract only language-independent files and files for the given language. By default all files are extracted.

To also skip language-independent files, combine this option with \fB\-\-language\-only\fP.
.TP
\fB\-\-language\-only\fP
Only extract files that are language-specific.

This option can be combined with \fB\-\-language\fP to only extract the files of a specific language.
.TP
\fB\-\-license\fP
Show license information.
.TP
\fB\-l\fP, \fB\-\-list\fP
List files contained in the installer but don't extract anything.

This action also enabled the \fB\-\-list\-sizes\fP action unless either \fB\-\-quiet\fP or \fB\-\-silent\fP is specified.

This option can be combined with \fB\-\-silent\fP to print only the names of the contained files (one per line) without additional syntax that would make consumption by other scripts harder.

The \fB\-\-list\fP action can be combined with \fB\-\-test\fP, \fB\-\-extract\fP, \fB\-\-list\-languages\fP and/or \fB\-\-gog\-game\-id\fP to display the names of the files as they are extracted even with \fB\-\-silent\fP.
.TP
\fB\-\-list\-checksums\fP
List checksums for files contained in the installer.

This option implies the \fB\-\-list\fP action and can be combined with the \fB\-\-list\-sizes\fP option to print both the size and checksum for each file.

With \fB\-\-silent\fP the file checksum will be printed at the start of the line (but after the file size if enabled with the \fB\-\-list\-sizes\fP option) followed by a space. Otherwise the checksum is printed after the file name.

The checksum type can be one of \fBAdler32\fP, \fBCRC32\fP, \fBMD5\fP or \fBSHA-1\fP and is printed in fron of the checksum hash followed by a space. \fBAdler32\fP and \fBCRC32\fP checksums are printed as "\fB0x\fP" followed by the 32-bit hexadecimal value.

Different files in the same installer can have different checksum types if GOG Galaxy file part reassembly is not disabled using the \fB\-\-no\-gog\-galaxy\fP option.
.TP
\fB\-\-list\-languages\fP
List languages supported by the installer.

This option can be combined with \fB\-\-silent\fP to print only the identifiers of the languages (one per line) followed by a space and then the language name, without additional syntax that would make consumption by other scripts harder.

The \fB\-\-list\-languages\fP action can be combined with \fB\-\-list\fP, \fB\-\-test\fP, \fB\-\-extract\fP and/or \fB\-\-gog\-game\-id\fP to display the available languages before doing anything else. If \fB\-\-silent\fP and \fB\-\-list\-languages\fP are combined with \fB\-\-list\fP and/or \fB\-\-gog\-game\-id\fP, the languages list will be terminated with an empty line and will precede both the game ID and files list.
.TP
\fB\-\-list\-sizes\fP
List uncompressed sizes for files contained in the installer.

This option implies the \fB\-\-list\fP action and can be combined with the \fB\-\-list\-checksums\fP option to print both the size and checksum for each file.

With \fB\-\-silent\fP the file size in bytes will be printed at the start of the line followed by a space. Otherwise the size is printed after the file name in a human-friendly format.
.TP
\fB\-L\fP, \fB\-\-lowercase\fP
Convert filenames stored in the installer to lower-case before extracting.
.TP
\fB\-d\fP, \fB\-\-output\-dir\fP \fIDIR\fP
Extract all files into the given directory. By default, \fBinnoextract\fP will extract all files to the current directory.

If the specified directory does not exist, it will be created. However, the parent directory must exist or extracting will fail.
.TP
\fB\-P\fP, \fB\-\-password \fIPASSWORD\fP
Specifies the password to decrypt encrypted files. The password is assumed to be encoded as UTF-8 and converted the internal according used in the installer as needed.

Use the \fB\-\-password-file\fP option to load the password from a file or standard input instead. This option cannot be combined with \fB\-\-password-file\fP.

If this password does not match the checksum stored in the installer, encrypted files will be skipped but unencrypted files will still be extracted. Use the \fB\-\-check\-password\fP option to abort processing entirely if the password is incorrect.
.TP
\fB\-\-password-file\fP \fIFILE\fP
Load a password form the specified file. Only the first line excluding the terminating carriage return and/or line break is used as the password. The password is assumed to be encoded as UTF-8 and converted the internal according used in the installer as needed.

If the special file name "\fB-\fP" is used, the password will be read from standard input.

Use the \fB\-\-password\fP option to specify the password on the command\-line instead. This option cannot be combined with \fB\-\-password\fP.

If this password does not match the checksum stored in the installer, encrypted files will be skipped but unencrypted files will still be extracted. Use the \fB\-\-check\-password\fP option to abort processing entirely if the password is incorrect.
.TP
\fB\-p\fP, \fB\-\-progress\fP[=\fIENABLE\fP]
By default \fBinnoextract\fP will try to detect if the terminal supports shell escape codes and enable or disable progress bar output accordingly. Pass \fB1\fP or \fBtrue\fP to \fB\-\-progress\fP to force progress bar output. Pass \fB0\fP or \fBfalse\fP to never show a progress bar.
.TP
\fB\-q\fP, \fB\-\-quiet\fP
Less verbose output.
.TP
\fB\-\-show\-password\fP
Show checksum \fB$c\fP and salt \fB$s\fP used for the password \fB$p\fP check as well as encoding of the password. The checksum is calculated from the salt concatenated with the password:

 \fB$c = hash($s . $p)\fP

With \fB\-\-silent\fP option, the checksum name and hash is printed on one line seperated buy a space followed by the salt encoded as hex bytes and password encoding on separate lines.

Checksum types can be \fBCRC32\fP, \fBMD5\fP or \fBSHA-1\fP although \fBCRC32\fP is not used in installers with encryption.

The password encoding is either \fBMS-ANSI\fP (Windows-1252) or \fBUTF16-LE\fP.

Use the \fB\-\-password\fP or \fB\-\-password\-file\fP option together with \fB\-\-check\-password\fP to check if a password matches this checksum.
.TP
\fB\-s\fP, \fB\-\-silent\fP
Don't output anything except errors and warnings unless explicitly requested and use a machine-readable output format.

This option can be combined with \fB\-\-list\fP to print only the names of the contained files (one per line) without additional syntax that would make consumption by other scripts harder.
.TP
\fB\-t\fP, \fB\-\-test\fP
Test archive integrity but don't write any output files.

This option can be combined with \fB\-\-extract\fP to abort on file checksum errors.
.TP
\fB\-T\fP, \fB\-\-timestamps\fP \fITZ\fP
Inno Setup installers can contain timestamps in both UTC and 'local' timezones.

The \fB\-\-timestamps\fP option specifies what timezone should be used to adjust these 'local' file times.

Valid values are those accepted by \fBtzset\fP in the \fBTZ\fP environment variable, except with the direction of the time offset reversed: both \fB\-T CET\fP and \fB\-T GMT+1\fP will (when DST is in effect) give the same result.

Besides timezones, two special values are accepted:

.RS
.HP
"\fBnone\fP"
Don't preserve file times for extracted files, both for UTC and 'local' timestamps. The file times wil be left the way the OS set them when creating the output files.
.HP
"\fBlocal\fP"
Use the system timezone for 'local' timestamps. This is the normal Inno Setup behavior, and can be used together with the \fBTZ\fP environment variable.
.RE
.IP

The default value for this option is \fBUTC\fP, causing innoextract to not adjust 'local' file times. File times marked as UTC in the Inno Setup file will never be adjusted no matter what \fB\-\-timestamps\fP is set to.
.TP
\fB\-v\fP, \fB\-\-version\fP
Print the \fBinnoextract\fP version number and supported Inno Setup versions.

If combined with the \fB\-\-silent\fP option, only the version \fInumber\fP is printed. Otherwise, the output will contain the name (innoextract) followed by the version number on the first line, and, unless the \fB\-\-quiet\fP options is specified, the range of suuported Inno Setup installer versions on the second line.
.TP
\fB\-\-no\-warn\-unused\fP
By default, innoextract will print a warning if it encounters \fI.bin\fP files that look like they could be part of the setup but are not used. This option disables that warning.
.SH PATH CONSTANTS
Paths in Inno Setup installers can contain constants (variable or code references) that are expanded at install time. innoextract expands all such constants to their name  and replaces unsafe characters with \fB$\fP. For exmaple \fB{app}\fP is expanded to \fBapp\fP while \fB{code:Example}\fP is expanded to \fBcode$Example\fP.

There is currently no way to configure this expansion except for disabling it with the \fB\-\-dump\fP option.
.SH EXIT VALUES
.PP
.IP \fB0\fP
Success
.IP \fB1\fP
Syntax or usage error
.IP \fB2+\fP
Broken or unsupported setup file, or input/output error
.SH LIMITATIONS
There is no support for extracting individual components and limited support for filtering by name.

Included scripts and checks are not executed.

The mapping from Inno Setup constants like the application directory to subdirectories is hard-coded.

Names for data slice/disk files in multi-file installers must follow the standard naming scheme.
.SH SEE ALSO
\fBcabextract\fP(1), \fBunar\fP(1), \fBunrar\fP(1), \fBunshield\fP(1), \fBtzset\fP(3)
.SH BUGS
.PP
Please report bugs to http://innoextract.constexpr.org/issues.
.SH CREDITS
.PP
\fBinnoextract\fP is distributed under the zlib/libpng license.  See the LICENSE file for details.
.PP
A website is available at http://constexpr.org/innoextract/.
.PP
This program uses the excellent lzma/xz decompression library written by Lasse Collin.
.SH AUTHOR
Daniel Scharrer (daniel@constexpr.org)


CHANGELOG

innoextract 1.7 (2018-06-12)
 - Added support for Inno Setup 5.6.0 installers
 - Added support for new GOG installers with GOG Galaxy file parts
 - Added support for encrypted installers with the --password (-P) and --password-file options
 - Added a --show-password option to print password check information
 - Added a --check-password option to abort if the provided password does not match the stored checksum
 - Added a --info (-i) convenience option to print information about the installer
 - Added a --list-sizes option to print file sizes even with --quiet or --silent
 - Added a --list-checksums option to print file checksums
 - Added a --data-version (-V) option to print the data version and exit
 - Added a --no-extract-unknown (-n) option to abort on unknown Inno Setup data versions
 - Fixed building in paths that contain regex expressions
 - Fixed case-sensitivity in parent directory when creating subdirectories
 - Fixed .bin slice file names used with Inno Setup versions older than 4.1.7
 - Fixed build with newer libc++ versions
 - Made loading of .bin slice files case-insensitive
 - The --test option can now be combined with --extract to abort on file checksum errors
 - Now compiles in C++17 mode if supported

innoextract 1.6 (2016-03-24)
 - Added support for Inno Setup 5.5.7 (and 5.5.8) installers
 - Added a --collisions=rename-all option
 - Changed --collisions=rename to omit the suffix for the file that would have been extracted with --collisions=overwrite instead of the first encountered file
 - Fixed @lang suffix sometimes missing for the first file with the --collisions=rename option
 - Fixed build error with CMake 3.5
 - Now compiles in C++14 mode if supported
 - Unsafe characters in special constant strings (ie ':' in {code:…}) are now replaced with '$'
 - Windows: Fixed error message if the source file could not be opened
 - Windows: Fixed progress bar flickering while printing extracted filenames
 - Windows binaries: Fixed crash on platforms without AVX support

innoextract 1.5 (2015-09-24)
 - Added support for Inno Setup 5.5.6 installers
 - Added support for a modified Inno Setup 5.5.0 variant
 - Added support for building without iconv (Windows-1252 and UTF-16LE only)
 - Added warnings for .bin files that are not part of the installer
 - Added a simple --include (-I) option to filter files thanks to Alexandre Detiste
 - Added a --list-languages option to list available languages
 - Added a --exclude-temp (-m) option to not extract temporary files
 - Added a --language-only option to skip language-independent files
 - Added a --collisions option to abort or rename files on collision
 - Added a --default-language option to prefer a language in case of file collisions
 - Added a --gog-game-id option to print the GOG.com game ID for Galaxy-ready installers
 - Added a --gog (-g) option to extract additional .bin files using unrar or unar
 - Fixed handling of spaces in the --data-dir option
 - Fixed an infinite loop with truncated LZMA streams
 - Fixed handling of forward slashes in stored file paths
 - Fixed size display for powers of 1024
 - Fixed loading headers if there are encrypted chunks
 - Fixed file collisions not being handled case-insensitively
 - Files will now be extracted into the same directory even if the stored case differs
 - Empty directories are now created correctly when extracting
 - Skipped files in encrypted chunks are now listed individually
 - Temporary files are now marked in file listings
 - Error summary is now written to stderr when using --quiet
 - Colors are now only enabled automatically if $TERM is set to something other than "dumb"
 - Improved error and warning messages
 - Build system improvements
 - Debug output can now be enabled separately from debug builds
 - Windows: Added support for using the Win32 API for string encoding conversion
 - Windows: Fixed unicode console output

innoextract 1.4-windows-r2 (2014-04-17)
 - This is a Windows-specific hotfix release, no other platforms are affected
 - Fixed running innoextract under Windows XP

innoextract 1.4-windows-r1 (2013-05-14)
 - This is a Windows-specific hotfix release, no other platforms are affected
 - Fixed a crash on startup under Windows
 - Reduced progress bar flickering under Windows
 - Fixed original console text color not being restored under Windows

innoextract 1.4 (2013-03-11)
 - Fixed build on non-Linux platforms with a separate libiconv (Windows™, macOS)
 - Fixed build on systems with non-standard iconv function prototypes (FreeBSD)
 - Fixed MSVC build
 - Fixed build with older glibc versions
 - Fixed issues with the progress bar in sandbox environments
 - Fixed string conversion on systems where libiconv defaults to big-endian variants
 - Fixed extracting very large installers with 32-bit innoextract builds
 - Improved handling of invalid encoded strings
 - Improved error messages when input or output files could not be opened
 - The --list command-line option can now combined with --test or --extract
 - The --version command-line option can now be modified with --quiet or --silent
 - Added color output and progress bar support for Windows™
 - Added support for Unicode filenames under Windows™
 - Added support for preserving timestamps of extracted files (enabled by default)
 - Added a --timestamps (-T) command-line options to control or disable file timestamps
 - Added an --output-dir (-d) command-line option to control where files are extracted
 - Added various CMake options to fine-tune the build process
 - Various bug fixes and tweaks

innoextract 1.3 (2012-07-03)
 - Fixed --quiet and --silent flags being ignored for some multi-file installers output
 - Now compiles in C++11 mode if supported
 - Added a warning when extracting unsupported setup data versions
 - Added support for Inno Setup 5.5.0 installers

innoextract 1.2 (2012-04-01)
 - Fixed compile errors with older versions of Boost or GCC.

src/InnoExtract.cpp

#include <fstream>
#include <iostream>
#include <iomanip>
#include <string>
#include <sstream>
#include <algorithm>
#include <cstring>
#include <vector>
#include <bitset>
#include <ctime>
#include <map>

#include <boost/shared_ptr.hpp>
#include <boost/foreach.hpp>
#include <boost/ref.hpp>
#include <boost/make_shared.hpp>
#include <boost/filesystem/path.hpp>
#include <boost/iostreams/copy.hpp>
#include <boost/date_time/posix_time/ptime.hpp>
#include <boost/date_time/posix_time/posix_time_types.hpp>

#include "version.hpp"

#include "cli/debug.hpp"

#include "loader/offsets.hpp"

#include "setup/data.hpp"
#include "setup/file.hpp"
#include "setup/info.hpp"
#include "setup/version.hpp"

#include "stream/chunk.hpp"
#include "stream/file.hpp"
#include "stream/slice.hpp"

#include "util/console.hpp"
#include "util/load.hpp"
#include "util/log.hpp"
#include "util/output.hpp"

using std::cout;
using std::string;
using std::endl;
using std::setw;
using std::setfill;

namespace io = boost::iostreams;
namespace fs = boost::filesystem;

int main(int argc, char * argv[]) {
	
	color::init();
	
	// logger::debug = true;
	
	if(argc <= 1) {
		cout << "usage: innoextract <Inno Setup installer>" << endl;
		return 1;
	}
	
	if(!strcmp(argv[1], "-v") || !strcmp(argv[1], "--version")) {
		cout << innoextract_version << endl;
		return 1;
	}
	
	std::ifstream ifs(argv[1], std::ios_base::in | std::ios_base::binary | std::ios_base::ate);
	if(!ifs.is_open()) {
		log_error << "error opening file \"" << argv[1] << '"';
		return 1;
	}
	
	loader::offsets offsets;
	offsets.load(ifs);
	
	cout << std::boolalpha;
	
	if(logger::debug) {
		print_offsets(offsets);
		cout << '\n';
	}
	
	ifs.seekg(offsets.header_offset);
	setup::info info;
	info.load(ifs);
	
	const std::string & name = info.header.app_versioned_name.empty()
	                           ? info.header.app_name : info.header.app_versioned_name;
	cout << "Extracting \"" << color::green << name << color::reset
	     << "\" - setup data version " << color::white << info.version << color::reset << std::endl;
	
	if(logger::debug) {
		cout << '\n';
		print_info(info);
	}
	
	cout << '\n';
	
	std::vector<std::vector<size_t> > files_for_location;
	files_for_location.resize(info.data_entries.size());
	for(size_t i = 0; i < info.files.size(); i++) {
		if(info.files[i].location < files_for_location.size()) {
			files_for_location[info.files[i].location].push_back(i);
		}
	}
	
	typedef std::map<stream::file, size_t> Files;
	typedef std::map<stream::chunk, Files> Chunks;
	Chunks chunks;
	for(size_t i = 0; i < info.data_entries.size(); i++) {
		setup::data_entry & location = info.data_entries[i];
		if(location.chunk.compression == stream::UnknownCompression) {
			location.chunk.compression = info.header.compression;
		}
		chunks[location.chunk][location.file] = i;
	}
	
	boost::shared_ptr<stream::slice_reader> slice_reader;
	
	if(offsets.data_offset) {
		slice_reader = boost::make_shared<stream::slice_reader>(argv[1], offsets.data_offset);
	} else {
		fs::path path(argv[1]);
		slice_reader = boost::make_shared<stream::slice_reader>(path.parent_path().string() + '/',
		                                                        path.stem().string(),
		                                                        info.header.slices_per_disk);
	}
	
	try {
	
	BOOST_FOREACH(const Chunks::value_type & chunk, chunks) {
		
		cout << "[starting " << chunk.first.compression << " chunk @ " << chunk.first.first_slice
		     << " + " << print_hex(offsets.data_offset) << " + " << print_hex(chunk.first.offset)
		     << ']' << std::endl;
		
		stream::chunk_reader::pointer chunk_source;
		chunk_source = stream::chunk_reader::get(*slice_reader, chunk.first);
		
		uint64_t offset = 0;
		
		BOOST_FOREACH(const Files::value_type & location, chunk.second) {
			const stream::file & file = location.first;
			
			if(file.offset < offset) {
				log_error << "bad offset";
				return 1;
			}
			
			if(file.offset > offset) {
				std::cout << "discarding " << print_bytes(file.offset - offset) << std::endl;
				discard(*chunk_source, file.offset - offset);
			}
			offset = file.offset + file.size;
			
			std::cout << "-> reading ";
			bool named = false;
			BOOST_FOREACH(size_t file_i, files_for_location[location.second]) {
				if(!info.files[file_i].destination.empty()) {
					std::cout << '"' << info.files[file_i].destination << '"';
					named = true;
					break;
				}
			}
			if(!named) {
				std::cout << "unnamed file";
			}
			std::cout << " @ " << print_hex(file.offset)
			          << " (" << print_bytes(file.size) << ')' << std::endl;
			
			crypto::checksum checksum;
			
			stream::file_reader::pointer file_source;
			file_source = stream::file_reader::get(*chunk_source, file, &checksum);
			
			BOOST_FOREACH(size_t file_i, files_for_location[location.second]) {
				if(!info.files[file_i].destination.empty()) {
					std::ofstream ofs(info.files[file_i].destination.c_str());
					
					char buffer[8192 * 10];
					
					float status = 0.f;
					uint64_t total = 0;
					
					std::ostringstream oss;
					float last_rate = 0;
					
					int64_t last_milliseconds = 0;
					
					boost::posix_time::ptime start(boost::posix_time::microsec_clock::universal_time());
					
					while(!file_source->eof()) {
						
						std::streamsize n = file_source->read(buffer, ARRAY_SIZE(buffer)).gcount();
						
						if(n > 0) {
							
							ofs.write(buffer, n);
							
							total += uint64_t(n);
							float new_status = float(size_t(1000.f * float(total) / float(file.size)))
							                   * (1 / 1000.f);
							if(status != new_status && new_status != 100.f) {
								
								boost::posix_time::ptime now(boost::posix_time::microsec_clock::universal_time());
								int64_t milliseconds = (now - start).total_milliseconds();
								
								if(milliseconds - last_milliseconds > 200) {
									last_milliseconds = milliseconds;
									
									if(total >= 10 * 1024 && milliseconds > 0) {
										float rate = 1000.f * float(total) / float(milliseconds);
										if(rate != last_rate) {
											last_rate = rate;
											oss.str(string()); // clear the buffer
											oss << std::right << std::fixed << std::setfill(' ') << std::setw(8)
											    << print_bytes(rate) << "/s";
										}
									}
									
									status = new_status;
									progress::show(status, oss.str());
								}
							}
						}
					}
					
					break; // TODO ...
				}
			}
			
			progress::clear();
			
			if(checksum != file.checksum) {
				log_warning << "checksum mismatch:";
				log_warning << "actual:   " << checksum;
				log_warning << "expected: " << file.checksum;
			}
		}
	}
	
	} catch(std::ios_base::failure e) {
		log_error << e.what();
	}
	
	std::cout << color::green << "Done" << color::reset << std::dec;
	
	if(logger::total_errors || logger::total_warnings) {
		std::cout << " with ";
		if(logger::total_errors) {
			std::cout << color::red << logger::total_errors << " errors" << color::reset;
		}
		if(logger::total_errors && logger::total_warnings) {
			std::cout << " and ";
		}
		if(logger::total_warnings) {
			std::cout << color::yellow << logger::total_warnings << " warnings" << color::reset;
		}
	}
	
	std::cout << '.' << std::endl;
	
	return 0;
}
